diff --git a/flags/Android.bp b/flags/Android.bp
index 9a2f9f2..982a56c 100644
--- a/flags/Android.bp
+++ b/flags/Android.bp
@@ -32,6 +32,7 @@ java_aconfig_library {
     defaults: ["framework-minus-apex-aconfig-java-defaults"],
     min_sdk_version: "35", // Make it 36 later
     apex_available: [
+        "//apex_available:platform",
         "com.android.nfcservices",
     ],
     visibility: [
